packages/concerto-core/lib/decoratormanager.js

```diff
@@ -253,6 +253,233 @@ class DecoratorManager {
         return newModelManager;
     }
 
+    /**
+     * Extracts all the decorator commands from all the models in modelManager
+     * @param {ModelManager} modelManager the input model manager
+     * @param {object} options - decorator models options
+     * @param {boolean} options.removeDecoratorsFromModel - flag to strip out decorators from models
+     * @param {string} options.locale - locale for extracted vocabulary set
+     * @returns {Object} a new model manager with the decorations removed and a list of extracted decorator jsons
+     */
+    static extractDecorators(modelManager,options) {
+        options = {
+            ...{
+                removeDecoratorsFromModel: false,
+                locale:'en'
+            },
+            ...options || {}
+        };
+        const ast = modelManager.getAst(true);
+        const decoratedAst = JSON.parse(JSON.stringify(ast));
+        let extractionDictionary={};
+        let processedModels = decoratedAst.models.map((model)=>{
+            if (model.decorators && model.decorators.length>0){
+                extractionDictionary=this.constructDCSDictionary(extractionDictionary,model.namespace, model.decorators,'','');
+                if (options.removeDecoratorsFromModel){
+                    model.decorators=undefined;
+                }
+            }
+            let processedDecl = model.declarations.map((decl)=>{
+                if (decl.decorators) {
+                    extractionDictionary=this.constructDCSDictionary(extractionDictionary,model.namespace,decl.decorators,decl.name,'');
+                }
+                if (options.removeDecoratorsFromModel){
+                    decl.decorators=undefined;
+                }
+                if (decl.properties) {
+                    let processedProperties = decl.properties.map((property) => {
+                        if (property.decorators){
+                            extractionDictionary=this.constructDCSDictionary(extractionDictionary,model.namespace, property.decorators,decl.name,property.name);
+                        }
+                        if (options.removeDecoratorsFromModel){
+                            property.decorators=undefined;
+                        }
+                        return property;
+                    });
+                    decl.properties=processedProperties;
+                }
+                return decl;
+            });
+            model.declarations=processedDecl;
+            return model;
+        });
```

Comment:
Please verify that this provides full coverage to all decorators. See the metamodel for reference, https://github.com/accordproject/concerto-metamodel/blob/main/lib/metamodel.cto

For example, does it also work for MapDeclarations and EnunDeclarations. 

A visitor pattern may be more appropriate here.

Reply:
Visitor Pattern may not work best here because I am modifying the ast for simplicity while removing decorators from source. 
The same approach (iterating on ast) was taken in applying the decorators as well. ( I am guessing it's because the APIS are read-only and we need to modify the ast to modify the model )
I have ensured it supports all types of declarations and corresponding props, including maps, scalars and enums. Have added a relevant test case as well.